tianshou/evaluation/rliable_evaluation_hl.py

```diff
@@ -61,61 +61,140 @@ class RLiableExperimentResult:
-                    f"but got {restored_test_data=}.",
+            try:
+                # TODO: fix
+                logger_factory = Experiment.from_directory(entry.path).logger_factory
+                logger_cls = type(logger_factory.create_logger(entry.path, entry.name, None))
+            # Usually this means from low-level API
+            except FileNotFoundError:
+                log.info(
+                    f"Could not find persisted experiment in {entry.path}, using default logger.",
                 )
+                logger_cls = TensorboardLogger
+
+            data = logger_cls.restore_logged_data(entry.path)
+            # TODO: align low-level and high-level dir structure. This is a hack!
+            if not data:
+                dirs = [
+                    d for d in os.listdir(entry.path) if os.path.isdir(os.path.join(entry.path, d))
+                ]
+                if len(dirs) != 1:
+                    raise ValueError(
+                        f"Could not restore data from {entry.path}, "
+                        f"expected either events or exactly one subdirectory, ",
+                    )
+                data = logger_cls.restore_logged_data(os.path.join(entry.path, dirs[0]))
+            if not data:
+                raise ValueError(f"Could not restore data from {entry.path}.")
+
+            if DataScope.TEST not in data or not data[DataScope.TEST]:
+                continue
+            restored_test_data = data[DataScope.TEST]
+            restored_train_data = data[DataScope.TRAIN]
+
+            assert isinstance(restored_test_data, dict)
+            assert isinstance(restored_train_data, dict)
+
+            for restored_data, scope in zip(
+                [restored_test_data, restored_train_data],
+                [DataScope.TEST, DataScope.TRAIN],
+                strict=True,
+            ):
+                if not isinstance(restored_data, dict):
+                    raise RuntimeError(
+                        f"Expected entry with key {scope} data to be a dictionary, "
+                        f"but got {restored_data=}.",
+                    )
             test_data = LoggedCollectStats.from_data_dict(restored_test_data)
+            train_data = LoggedCollectStats.from_data_dict(restored_train_data)
 
-            if test_data.returns_stat is None:
-                continue
-            test_episode_returns.append(test_data.returns_stat.mean)
-            env_step_at_test = test_data.env_step
+            if test_data.returns_stat is not None:
+                test_episode_returns.append(test_data.returns_stat.mean)
+                env_step_at_test = test_data.env_step
+
+            if train_data.returns_stat is not None:
+                train_episode_returns.append(train_data.returns_stat.mean)
+                env_step_at_train = train_data.env_step
 
+        test_data_found = True
+        train_data_found = True
         if not test_episode_returns or env_step_at_test is None:
```

Comment:
We probably need another mechanism here as well. In the unfortunate event of all but the last run having recorded data, `env_step_at_test` may be None and the RuntimeError below is raised.

Reply:
I feel like we should rather improve the logging and stats-saving mechanism to ensure this never happens instead of doing more gymnastics here. Would that be possible? We can do that in a separate issue/PR.